[wip] update ns1/ns1-go to v2

josephholsten opened the pull request:
we changed a bunch, so this will probably take a bit

josephholsten committed:
vendor: update ns1/ns1-go to v2

josephholsten committed:
nsone: update to ns1/ns1-go.v2

Notable changes:
- update to import gopkg.in/ns1/ns1-go.v2/rest
- update to use rest.NewClient()
- update to use rest.*Service types for CRUD calls
- update to use gopkg.in/ns1/ns1-go.v2/rest/model/* packages for structs
- dramatic rewriting to support data.Meta

pashap commented on builtin/providers/nsone/resource_apikey.go:
```diff
@@ -118,64 +120,64 @@ func resourceDataToPermissions(d *schema.ResourceData) nsone.PermissionsMap {
-	client := meta.(*nsone.APIClient)
-	mj := nsone.Apikey{}
+	client := meta.(*nsone.Client)
+	mj := account.APIKey{}
```
i think `mj` is short for monitoring job. not a big deal at all, but maybe this should be `a` or `ak` for apikey. but again, nbd

josephholsten replied:
resolved

pashap commented on builtin/providers/nsone/resource_apikey.go:
```diff
@@ -118,64 +120,64 @@ func resourceDataToPermissions(d *schema.ResourceData) nsone.PermissionsMap {
-	client := meta.(*nsone.APIClient)
-	mj, err := client.GetApikey(d.Id())
+	client := meta.(*nsone.Client)
+	mj, _, err := client.APIKeys.Get(d.Id())
```
same here regarding `mj`

josephholsten replied:
resolved

pashap commented on builtin/providers/nsone/resource_record.go:
```diff
@@ -264,17 +269,25 @@ func recordToResourceData(d *schema.ResourceData, r *nsone.Record) error {
 			newRegion["name"] = regionName
-			if len(region.Meta.GeoRegion) > 0 {
-				newRegion["georegion"] = region.Meta.GeoRegion[0]
+			// TODO: support as FeedPtr
```
Could we do a switch on `region.Meta.Georegion.(type)` and check if either `data.FeedPtr` or `[]string`?

pashap replied:
That seems nasty to have to do to every metadata field tho...

but im not sure of a cleaner way

josephholsten replied:
that's exactly what I'm working on now

pashap commented on builtin/providers/nsone/provider.go:
```diff
@@ -32,9 +35,10 @@ func Provider() terraform.ResourceProvider {
-	n.RateLimitStrategySleep()
+	httpClient := &http.Client{}
+	n := nsone.NewClient(httpClient, nsone.SetAPIKey(d.Get("apikey").(string)))
+	// FIXME: n.Debug()
```
i wasnt sure what to do here either. `v1` prints requests and responses when debug is enabled.

we can add a logger that logs every request with the following if that is what users would want.

```
    // Adds logging to each http request.
    doer := nsone.Decorate(httpClient, nsone.Logging(log.New(os.Stdout, "", log.LstdFlags)))
    client := nsone.NewClient(doer, nsone.SetAPIKey(k))
```

josephholsten replied:
this should also be tied to terraform's debug state

pashap replied:
Sorry, trying to find docs on terraform debug state now. We can add whatever logs to the requests/responses that we want with the decorator. my guess is that it should be logging both requests and responses. is their any documentation on the debug state?

pashap commented on builtin/providers/nsone/provider.go:
```diff
@@ -32,9 +35,10 @@ func Provider() terraform.ResourceProvider {
+	httpClient := &http.Client{}
+	n := nsone.NewClient(httpClient, nsone.SetAPIKey(d.Get("apikey").(string)))
+	// FIXME: n.Debug()
+	// n.RateLimitStrategySleep()
```
This should still be applicable for `v2`. if users would like to have it

josephholsten replied:
I got a divide by zero panic that I haven't had a change to investigate. I absolutely want to re-enable it.

pashap replied:
Ah, im going to check that out now

pashap replied:
@josephholsten I made a PR for the fix: https://github.com/ns1/ns1-go/pull/24

The client was applying the rate limit func before checking the actual response code. I was able to reproduce divide by zero when i didnt set the apikey as an env var and received a 401(without rate limiting headers).

pashap replied:
i merged in the rate limiting fix, so this should work now

josephholsten commented on builtin/providers/nsone/resource_datafeed.go:
```diff
@@ -33,28 +35,29 @@ func dataFeedResource() *schema.Resource {
-	df.Name = d.Get("name").(string)
-	config := make(map[string]string)
+func resourceDataToDataFeed(d *schema.ResourceData) *data.Feed {
+	config := make(data.Config)
```
- [ ] investigate using the `data.NewFeed()` constructor

josephholsten commented:
current failing tests:
```
=== RUN   TestAccMonitoringJob_updated
--- FAIL: TestAccMonitoringJob_updated (0.54s)
        testing.go:172: Step 0 error: Check failed: Check 3/3 error: Bad value : 0
=== RUN   TestAccRecord_basic
--- FAIL: TestAccRecord_basic (2.47s)
        testing.go:233: Error destroying resource! WARNING: Dangling resources
                may exist. The full state and error is shown below.

                Error: Check failed: Record still exists: &dns.Record{Meta:(*data.Meta)(nil), ID:"", Zone:"", Domain:"", Type:"", Link:"", TTL:0, UseClientSubnet:false, Answers:[]*dns.Answer(nil), Filters:[]*filter.Filter(nil), Regions:data.Regions(nil)}

                State: <no state>
=== RUN   TestAccRecord_updated
--- FAIL: TestAccRecord_updated (1.69s)
        testing.go:233: Error destroying resource! WARNING: Dangling resources
                may exist. The full state and error is shown below.

                Error: Check failed: Record still exists: &dns.Record{Meta:(*data.Meta)(nil), ID:"", Zone:"", Domain:"", Type:"", Link:"", TTL:0, UseClientSubnet:false, Answers:[]*dns.Answer(nil), Filters:[]*filter.Filter(nil), Regions:data.Regions(nil)}

                State: <no state>
```

josephholsten committed:
ultradns: simplify hashLimits()

Limits blocks only have the "name" attribute as their primary key, so
hashLimits() needn't use a buffer to concatenate.

josephholsten commented:
fixed `TestAccMonitoringJob_updated`

Still broken:

- `TestAccRecord_basic`
- `TestAccRecord_updated`

josephholsten commented:
OK, all tests are passing again

jmasseo reviewed (approved):
"LGTM"
